node_modules/oae-authz/lib/permissions.js

```diff
@@ -0,0 +1,728 @@
+ * @param  {String}             role                        The role in which to share
+ * @param  {Function}           callback                    Standard callback function
+ * @param  {Object}             callback.err                Thrown if the permission check fails
+ * @param  {MemberChangeInfo}   callback.memberChangeInfo   The members change info that should be applied to complete the share
+ * @param  {EmailChangeInfo}    callback.emailChangeInfo    The email change info that should be applied to perform the necessary invitations
+ */
+var canShare = module.exports.canShare = function(ctx, resource, targets, role, callback) {
+    targets = targets || [];
+
+    // Determine what access, if any, the user in context has on the resource
+    resolveEffectivePermissions(ctx, resource, function(err, permissions) {
+        if (err) {
+            return callback(err);
+        } else if (!permissions.canShare) {
+            return callback({'code': 401, 'msg': 'The current user does not have access to share this resource'});
+        }
+
+        // Apply the share role to all the targets
+        _.each(targets, function(target) {
+            target.role = role;
+        });
+
+        // Ensure that no privacy is being violated by sharing with these targets
+        _validateRoleChanges(ctx, resource, targets, {'promoteOnly': true}, function(err, memberChangeInfo, emailChangeInfo) {
+            if (err) {
+                return callback(err);
+            } else if (permissions.canManage) {
+                // If we can manage the resource we don't need to check that the user in context can
+                // extend the explicit access of the resource outside the set visibility
+                return callback(null, memberChangeInfo, emailChangeInfo);
+            }
+
+            // The user is not a manager, so we should ensure that whomever they are adding as a
+            // share target already has implicit access
+            var addedPrincipals = _.chain(emailChangeInfo.emails.added)
+                .map(_emailToResource)
+                .union(memberChangeInfo.members.added)
+                .value();
+            if (_.isEmpty(addedPrincipals)) {
+                // If there are no actual changes to make, then this is basically an empty share
+                // which should be fine
+                return callback(null, memberChangeInfo, emailChangeInfo);
+            }
+
+            // Since we are not a manager, for each principal we are sharing with, we need to make
+            // sure all users already had implicit access to the resource (e.g., a non-manager can't
+            // share a loggedin resource with a public user from another tenant because that would
+            // violate the visibility that managers applied to the resource)
+            var invalidPrincipals = {};
+            var _done = _.chain(addedPrincipals).size().after(function() {
```

Comment:
I tend to default to using underscore where possible to access properties of collections/options as it will gracefully handle "empty" / "null" scenarios without exploding. However in this case, an empty collection would actually result in the request hanging, so it's probably better to have it explode if the collection isn't specified...